Initial implementation of OSC 7

skyline75489 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

This is a very early implementation of OSC 7 in WIndows Terminal 

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [X] Closes #3158 
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

skyline75489 commented:
I'll add the explanation here to make the PR description clean.

So I've went over basically every comments in #3158 . As a daily terminal users and former contributor of this repo, I understand most of the struggles from both the terminal users' side and WT developers' side. I decided to come up with a minimal, reliable and working approach, by adding the following constraints:

* Terminal **does not** probe the shell process running inside, like ever. 
* OSC 7 handling is strictly isolated between different **profiles**. 
* Terminal **does not** probe the type of the profile being used. I know that "source" being `Windows.Terminal.Wsl` is a strong sign that indicates the profile is a WSL one. But no. Don't do this.
* Terminal **does not** probe the type of the actual URI passed in by OSC 7. I know there's way to determine whether a URI is likely to be a Windows one or Linux one. But no. Don't do this.

With this constraints, I managed to make this PR. It's very simple. See for yourself. The diff contains merely 100 lines. To be honest I don't think it will be a very large PR even after polished, assuming that this approach is approved by the Console team and the community in the end.

I really hope that this particular PR itself can fully fixes #3158. However, with the terminal doing only the very basic stuff, the burden moves to the shells. I've managed to make both PowerShell and Bash/Zsh on WSL working. But I'm out of luck with CMD and Git Bash because I could not find a way to produce the correct sequences. I've searched many things before I could made a working script. Due to possible licensing issue, I'll link my scripts below and assume that it will not be included in this repo:

* [PowerShell OSC 7 Gist](https://gist.github.com/skyline75489/480d036db8ae9069b7009377e6eebb79)
* [Bash/Zsh on WSL OSC 7 Gist](https://gist.github.com/skyline75489/d655aede4c729eff178a1c0bfd10f622)

There's several issues remaining to be discussed/solved:

* CMD/Git Bash/etc not working. Some of them may not be possible
* Shells other than Bash/Zsh not tested. 
* Tab with multi-pane is not handled. I'm not a multi-pane user so I'm not sure what the expected behaviour is.
* When there are more than one tab (same profile) calling OSC 7, which URI should be used? And when should the URI be cleared? 

Aside from the known issues and other possible caveats, I'm having a blast with this. `Duplicate Tab` does exactly what you think it should do. The working directory is being successfully transferred between old tabs and new tabs. When using WSL in WT, it just reminds you of a normal Linux terminal. No hassle. No pain. Nothing.

Thanks everyone's participation in #3158 . Now let's see where this will take us.

skyline75489 commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -2372,6 +2372,16 @@ bool AdaptDispatch::EndHyperlink()
+{
+    if (_pConApi->IsConsolePty())
+    {
+        return false;
```
Well conhost does not have tabs. So this should be OK.

j4james replied:
If this isn't going to be implemented in conhost, then the method shouldn't be added to `AdaptDispatch` at all. Just let it fall back to the base implementation that returns false.

skyline75489 replied:
`ITermDispatch` is a virtual class. I can not fall back to the base implementation for this one.

j4james replied:
Sorry. My mistake. For some reason I assumed it derived from `TermDispatch` the same way `TerminalDispatch` does. It probably should, but that's not your problem.

skyline75489 commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -664,7 +665,29 @@ namespace winrt::TerminalApp::implementation
+                if (profileGuid == focuesdProfileGuid)
+                {
+                    auto workingDirectory = tabImpl->GetActiveTerminalControl().WorkingDirectory();
+                    auto validWorkingDirectory = !workingDirectory.empty() && ::PathFileExists(workingDirectory.data());
```
The validation is very important. Without this PowerShell and WSL will refuse to spawn when there's invalid path.

skyline75489 replied:
Surprisely enough, if you open a PowerShell tab and run `wsl.exe`, the path `\\wsl$\Ubuntu\path` sent by WSL can also be used by PowerShell. It will correctly enter the WSL working directory. I remember seeing this on some article about WSL and Windows filesystem interoperablity. I just didn't not except that it works so seamlessly.

And I think this only works on Windows 2004 and WSL 2. Correct me if I'm wrong.

WSLUser commented:
> Pure SSH profile does not work. It straight up ignores the current directory setting.

I think it's kinda important to get this one working. If I have a tab open where I have used ssh to connect to a host and I want to duplicate that session on another tab, we need to get this working. It should allow for a password prompt (because of ssh.exe)  if you don't have an ssh key or you configured a password on your ssh key but if you do have a password-less ssh key (the most common scenario), this should be a seamless experience.

LuanVSO commented:
[cmd OSC7](https://gist.github.com/LuanVSO/4804cadb2ce42676d76ed69390a60b0e)

WSLUser commented:
> I feel that it may require some modification in the OpenSSH-Win32 itself

If that is the case, I hope you could provide them with the needed info to get it working and/or provide them a PR to address the issue.

DHowett commented:
> > Pure SSH profile does not work. It straight up ignores the current directory setting. (@skyline75489)
>
> I think it's kinda important to get this one working. (@WSLUser)

I'm not exaggerating when I say this. Do not interpret this as a joke or as hyperbole.

This is **impossible** in the general case.

Unless you wish for Windows Terminal to take a dependency on the exact remote shell in use (you do not), you cannot propagate the working directory over ssh. There is no protocol feature in ssh--specifically, the part that handles remoting ptys and launching shells--that supports transmitting a working directory.

TBBle commented:
I think @zadjii-msft's approach of using the current control's working directory would allow removing the strict profile isolation. This would be nice, because as well as "duplicate tab", the other use-case I have for this functionality would be start a "git bash" session in the same directory my current PowerShell session is in. I'd be fine if this was a non-default behaviour for 'New tab for Profile X' (Ctrl-Shift-Alt-3 would still be faster than Ctrl-Shift-3, *beat*, `cd /d/project/repos/whateverIhappenedToBeWorkingOn` as I do often enough now).

TBBle commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -1628,6 +1634,24 @@ bool OutputStateMachineEngine::_ParseHyperlink(const std::wstring_view string,
+        return false;
+    }
+
+    uri = std::wstring(string.substr(7, string.length() - 7));
```
Is the plan to later add hostname validation, so we can filter-out SSH-sourced, Docker-sourced, and similar remote paths? That would eliminate a source of false-positives for the 'Is Valid Path' test, assuming existing code that generates OSC 7 is including the hostname. Hostname is described as "should" in the [Terminal.app specification](https://github.com/microsoft/terminal/issues/3158#issuecomment-578550385).

skyline75489 replied:
As much as I want to do this, there’s certain difficulty. First for my WSL distro, there’s no HOSTNAME in the env. There’s only HOST which is the computer name of the Windows OS. Similar for PowerShell. The concept of HOSTNAME does not to fit in the world of Windows. 

In the *nix world the path is interchangeable between the local machine and remote machine. However if you’re sshing to a Linux machine on Windows, the paths on the remote machine are guaranteed to be invalid on the local machine. My bash/zsh scripts actually make great use of the nature. The two types of the URIs it can generate are “c:\path” and “\\wsl$\path”. None of them is a valid *nix path. Pretty ironic now I put it this way. But it sure works. In the Windows world I can not find a good place for HOSTNAME.

TBBle replied:
Weird. I just checked my WSL2 Ubuntu, and I have `$HOSTNAME` but not `$HOST`. Both Git Bash and MSYS2 also have `$HOSTNAME`. In the Windows environment, it's `$COMPUTERNAME` (`$env:COMPUTERNAME` in PowerShell), and that env-var is visible in Git Bash and MSYS2 as well.

Either way, it's up to the person enabing OSC 7 to ensure they enable it to generate a hostname.

It would also be reasonable to accept hostname "" as "trust me, it's localhost" for simplicity. The main need here is to filter _out_ paths coming from remote machines, whether it's bash over SSH or PowerShell remoting, without relying on the remote host generating paths that are invalid locally.

skyline75489 replied:
Personally I’d like to stick to the “do not probe the path passed in” policy because it gives us more flexibility without bringing too much complexity to the terminal itself. Of course I’m no expert at VT, or terminal, or anything really. If there’s a better way most people can agree on, I’m totally cool with changing it.

TBBle replied:
I've just realised this is a bug: It's trimming off 7 characters, `file://`, but in fact the path starts after `file://<hostname>/`. So this as-is won't work with OSC 7 commands that work according to the Terminal.app help text, nor the [formerly-proposed spec at terminal-wg](https://gitlab.freedesktop.org/terminal-wg/specifications/-/blob/5f6bd72258b3d62e9cbc01908c32783d025f2893/proposals/working-path-notifications.md).

I don't consider it 'probing' the path to be correctly separating the part before the path, and since that will give you the shell-provided hostname anyway, ignoring it if it isn't "" or `$COMPUTERNAME` seems reasonable to me (and is what's described in the Terminal.app help text and the proposed spec)

skyline75489 replied:
Yes I intended to ignore anything about HOSTNAME, because as I said in
other comments, I can not find a meaningful thing to do with it. I wasn’t really trying to follow the spec when I wrote this. But you’re right. Following the spec helps the community in some way. 

Separating the HOSTNAME of course does not count as probing. But “filtering out SSH-sourced, Docker-sourced, and similar remote paths” based on that is something I chose to avoid.

TBBle replied:
That's why the spec was "Use hostname to recognise remote paths". It doesn't say you need to filter-out using that knowledge; and now that I know that `file://wsl$/Ubuntu` is the correct URL for the WSL mount, I don't think we can actually _ignore_ such paths here after all, and will have to rely on the 'is valid path' check when using it value.

skyline75489 replied:
Sadly `startingDirectory` does not accept `file://`. We just have to live with it.

skyline75489 commented:
Git bash does not recognize VT control sequences in WT the same way as it does in its own mintty. I don’t have a clue why. Theoretically it should work just fine if the control sequences can be generated.

DHowett commented:
> Git bash does not recognize VT control sequences

Cygwin versions < 3.1 (and therefore msys versions based on them) did not support `ENABLE_VIRTUAL_TERMINAL_PROCESSING` on Windows 10+, even though it's been out for 5 years.

They implemented their own VT parser.

skyline75489 commented:
After my experiments so far, there’s only one thing stopping me from lifting the same profile constraint - CMD can not recognize UNC path (\\wsl$\) generated by WSL CMD will fall back to c:\somewhere I don’t remember. Aside from this, I’m OK with cross profile OSC 7 handling. As i mentioned in other comment, as long as the URI is strictly Windows-specific, all shells that respond to CWD setting should work (again not sure if Git bash works).

TBBle commented:
[Git for Windows 2.27](https://github.com/git-for-windows/build-extra/blob/abf5904fa1878099ff069eac0c997ef95a155c7d/ReleaseNotes.md#changes-since-git-for-windows-v2262-april-20th-2020) upgraded to a Cygwin 3.1-based MSYS2.

> This release comes with a Git Bash that optionally uses [Windows-native pseudo consoles](https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/).

It's off by default, but unless running the installer in silent mode, it should prompt to enable it during install/upgrade.

I thought I had told Git for Windows to notify me of updates, but I was still using 2.24.0, and hadn't thought to go check for an update manually until now.

TBBle commented:
CMD wouldn't work with any UNC patch, without using pushd to get a temporary drive letter for it. So I don't think that's a problem for Windows Terminal here, and more than it would be a WT problem if one of the user's shells is configured to ignore the working directory and immediately `cd` somewhere else. Or... was a remote shell that _couldn't_ rationally honour the working directory.

j4james commented:
Now that I've had a chance to see how this is working, I'm afraid I wouldn't consider it an acceptable solution. It may solve the problem of opening a new tab in the current directory, but it's not a standard `OSC 7` implementation (at least as I understand it).

If a Linux shell has built-in support for `OSC 7`, and my current directory is `/home/james`, then the uri it's going to set will be something like `file://localhost/home/james`, and definitely not `file://\\\\wsl$\Ubuntu/home/james`. If we only understand the latter, and not the former, then we don't support `OSC 7`.

Similarly, if I have manually setup my shell to use a uri of the form `\\wsl$\Ubuntu\...` to get it to work in Windows Terminal, then that likely means it will no longer work in other terminals that actually support the standard path syntax. We've essentially taken over an existing sequence and repurposed it in a way that breaks other software - that's not good.

Bottom line: if we can't support `OSC 7` in a standard way, then I think we shouldn't be using it, and should just come up with our own proprietary sequence.

skyline75489 commented:
I thought about the difference between my current implementation and the standard. Here's my two cents. 

First when OSC 7 was first seen on Terminal.app, it's more like a proprietary escape code that for macOS exclusively. When terminal-wg decided to finalize a standard, they chose to use the [Freedesktop file URI spec](https://web.archive.org/web/20200705133622/https://edeproject.org/spec/file-uri-spec.txt)
 . This is a deviation from the Terminal.app one, even though in practice the difference should not really matter. The thing is, people would naturally expect things to work exactly the same between local macOS/Linux machine and remote Linux machines. That's of course because they're all *nix after all. So everything is good.

But here with WIndows Terminal that's not the case. Windows and *nix are fundamentally and historically different in so many ways. I don't think people would actually expect things to work exactly the same way between a local Windows machine and remote Linux machines. So I thinks it's perfectly OK to "have manually setup my shell to use a uri of the form \\wsl$\Ubuntu\... to get it to work in Windows Terminal, then that likely means it will no longer work in other terminals that actually support the standard path syntax". Because WSL is not and never will be a "real" Linux environment. Same for PowerShell/Git Bash/CMD/etc. The Windows Terminal & related shell environment are born different from the outside *nix world. I don't like it, but it's the world we live in. 

Point is, the OSC 7 in WT and related shell environments are destined to be exclusively working on Windows OS only. And I don't think people would be frustrated about it. The beauty of OSC 7 is, I think, that the methodology itself is platform-independent. That does not necessarily mean the actual implementation should be platform-independent. At least on Windows I don't see that's possible.

j4james commented:
> Because WSL is not and never will be a "real" Linux environment.

I don't think that's an accurate statement. Support for [GUI Linux apps running on WSL](https://devblogs.microsoft.com/commandline/the-windows-subsystem-for-linux-build-2020-summary/#wsl-gui) is definitely on the roadmap. Once that's the case, you could be running Gnome Terminal and Windows Terminal side by side accessing the same WSL subsystem.

But even ignoring that, there are already third party terminals (like Mintty) that can connect to WSL, and that support the standard `OSC 7` sequence. If we encourage people to ignore that standard and use a proprietary WT syntax instead, that will end up breaking the functionality in other terminals. This is exactly what Microsoft did with Internet Explorer back when it was the dominant browser, and that did not go down well.

If you don't want to follow the standard, that's fine - just invent your own sequence instead. There is absolutely no reason to use `OSC 7` in an incompatible way unless you're deliberately trying to break other software.

skyline75489 commented:
Honestly I can’t say that being compatible with current OSC 7 standard is technically not possible. Move the path handling logic from the zshrc to WT, find a way to detect WSL profile, put them together and it should work the same. It’s just I don’t really want this kind of “dirty” logic inside WT, just from my point of view. 

I don’t make the call here. And I’m willing to hear from everyone. So yeah. Not following the standard outside is not ideal, I admit.

TBBle commented:
I feel that "detect that it was WSL, identify which instance, and turn that WSL instance-internal `/path` into `//wsl$/<instanceName>/path`" is going to be only marginally less-awful than "Just scan the various possible CWD internal objects in the process control block", and hence not a reasonable result.

I agree that we should support a standard OSC 7, not introduce a new variant.

That said, I'd be curious to know how mintty running on Windows would handle OSC 7 from WSL, whether it expects `OSC 7 file://hostname/home/myaccount/`, or if it would expect `OSC 7 file://wsl$/Ubuntu/home/myaccount`.

If it's the latter, we already need OSC 7 to report Windows-resolvable `file://` URLs. If it's the former, it can't possibly _use_ that URL, since it isn't itself running in the WSL environment.

In the way OSC 7 has been described and implemented in Terminal.app and terminal-wg, at least, it assumes the `file://` URL is sensible for where the shell is running, and the shell doesn't know where the terminal is running, so it can't be expected to adjust the `file://` URL to suit it, e.g. it won't contain `wsl$` as the hostname or have the WSL distribution name prepended as the first path segment, which is what would be needed for a 'windows-local' `file://` URL.

Which means that OSC 7 for WSL is only slightly more rational than OSC 7 for ssh: Useful for showing in a titlebar, for example, but you can't carry it between sessions without much more intrusive logic to transform the 'remote' URL into a 'local' URL, and apply it. The only difference is that WSL has a feasible 'apply it', but the 'transform it' step is still going to be a sticking point.

skyline75489 commented:
Ok let me sort this out. First, if we are going to support OSC 7 strictly by standard, there has to be some kind of transformation applied to the URI to make it Window-local friendly, so that WT can use it without too much hassle. Second, using wsl$\ubuntu instead of localhost will damage other terminals who assume they’ll receive “localhost”. 

Honestly I’m OK with tweaking the zshrc to check if there’s WT_SESSION and change the HOSTNAME accordingly. It’s a WSL specific thing but with the check it will not break other software. If we insist to make it working “out of the box”, that is without changing anything to WSL specific, we are back to the dilemma discussed before.

j4james commented:
> there has to be some kind of transformation applied to the URI to make it Window-local friendly

Just to be clear, this transformation has to happen in the terminal itself, and not in the shell. Remember that some shells and applications will have the `OSC 7` functionality built in, so Windows Terminal needs to understand whatever syntax they send to us - we're not necessarily going to have any control of the format.

> using wsl$\ubuntu instead of localhost

Note that the `wsl$\ubuntu` portion of a WSL URI is not the host name - it's the start of a UNC path. A correctly formatted URI with a UNC path would be something like `file://localhost/\\wsl$\Ubuntu\home`. With the backslashes replaced, it would be `file://localhost///wsl$/Ubuntu/home` or `file://///wsl$/Ubuntu/home` without the hostname. You can get away with dropping some slashes, but that's the canonical format as far as I'm aware.

That said, this shouldn't be a concern to us when parsing a standard WSL path, but it could be something we need to handle in a cmd or powershell profile. Ideally we'd be using a URI library for the parsing, though, so it's probably best handled in the dispatch class (which would have access to the Foundation libraries) rather than the state machine.

skyline75489 commented:
I've updated the PR and related gists accordingly. The spec is now following `file://hostname/path`. But the hostname is filtered out before being passed to `workingDirectory`. This way we are not breaking compatibility with other OSC implementations. 

Thanks to @LuanVSO, CMD OSC 7 now works. Thanks to @TBBle , Git bash with VT enabled also works. Remaining things that do not work:

* UNC paths are not recognized by CMD and will fallback to `C:\Windows`.
* Pure MSYS paths are not recognized by other shells.

Thanks @j4james for the profound insights.

skyline75489 commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -1628,6 +1634,34 @@ bool OutputStateMachineEngine::_ParseHyperlink(const std::wstring_view string,
+
+    // Hostname is ignored.
+    current = nextSlash + 1;
+    uri = std::wstring(string.substr(current, std::wstring::npos));
```
This is just an early prototype. The actual implementation would require serious URL decoding and related things.

TBBle commented:
> A correctly formatted URI with a UNC path would be something like `file://localhost/\\wsl$\Ubuntu\home`.

That's not what I'm getting from [Wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme#Windows_2) (which references back to [an MSDN blog post from 2006](https://docs.microsoft.com/en-us/archive/blogs/ie/file-uris-in-windows)), nor from testing in Chrome, they both agree a correctly-formatted `file://` URL for a UNC path looks like `file://wsl$/Ubuntu/home`.

Wikipedia lists `file:////wsl$/Ubuntu/home` as a possible legacy format, but the blog post source for that is gone, although it's presumably still on the Wayback machine if that matters. Chrome accepted that, but not `file://localhost//wsl$/Ubuntu/home`.

skyline75489 commented:
The native Windows file:// URI does not work for CWD setting, making it kinda useless in this situation. So my approach is using only the path in the terminal-wg spec. Like I said in other comments,  I failed to find a good use for HOSTNAME in the Windows environment. I feel like I’m stealing the Freedesktop URI format and use it in a Windows-specific way. But I’m following the spec faithfully, at least on the interface side.

j4james commented:
> That's not what I'm getting from [Wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme#Windows_2) (which references back to [an MSDN blog post from 2006](https://docs.microsoft.com/en-us/archive/blogs/ie/file-uris-in-windows)), nor from testing in Chrome, they both agree a correctly-formatted ("healthy") `file://` URL for a UNC path looks like `file://wsl$/Ubuntu/home`.

Chrome is kind of the odd one out on this. IE, Edge (the original), Firefox, cmd.exe (using start), and Windows Explorer all accept the formats I mentioned. Chrome does actually support the last of the three, though, and that's what Firefox considers canonical (i.e. it reformats the others to that). What I was trying to show with the `file://localhost/\\wsl$\Ubuntu\home` syntax was where the 5 slashes come from, and where the hostname would fit if you needed to include it.

It's interesting that the MSDN blog post is claiming 2 slashes as the correct format, though. It makes sense for a network share, but not so much for any other UNC path. And there's no way to use that format to indicate that you're referring to a UNC path (that isn't a network share) on a different machine.

But ignoring UNC paths for the moment, in a WSL shell the requirements are simple. We need to support whatever format the other Linux and macOS terminals support, and the format that existing shells generate, and that's essentially `file://$HOSTNAME$PWD` (with appropriate escaping). When it comes to Windows shells, we've probably got more leeway to decide the format ourselves, but it makes sense to follow the same pattern. So in a cmd shell I'd expect something like `file://%USERDOMAIN%/$p` (using `prompt` syntax).

In both those cases UNC paths shouldn't really be an issue (they're not going to show up in a Linux $PWD and can't be used as the current directory in a cmd shell). So the only place we'd really need to worry about them is in powershell, and even then that's probably an edge case we can ignore if it's problematic.

j4james commented:
>  I failed to find a good use for HOSTNAME in the Windows environment.

The point of the host name is to distinguish between a path on a remote machine (which should be ignored) and a local one (which should be useable as the start path for the new tab). I don't think it's essential that we perform that validation - I'm not certain whether other terminals bother - but that's what the hostname is there for. 

I'm not sure why you think it wouldn't be applicable in Windows, but I assumed we'd just look for a match between the given hostname and one or more of the `GetComputerNameEx` names.

TBBle commented:
Firefox canonicalises to a different format that IE and Explorer canonicalise to. Deliberately so, according to that Firefox bug I linked, because the Firefox developers didn't like the way that Windows defined the `file://` URL. In a UNC path, the first thing after `\\` is the hostname, so `\\hostname\...` => `file://hostname/...` is the _most obvious_ transform.

It just so happens that non-Windows platforms can't usually do anything with that hostname, except filter the request, while Windows has a default mechanism for accessing remote machines through local (UNC) paths, and uses it for `file://` URLs.

As noted, the underlying Windows handler (urlmon, I think?) can _also_ deal with a `file://` URL of `file://localhost///hostname/...`, but that seems a weird choice to canonicalise to, because now we're packing a hostname into the path, and just hard-coding "localhost" into the hostname. And from my brief testing, this "legacy" format only works with _literal_ `file://localhost/` or literal `file:///`, which matches this clause from the RFC:

> As a special case, <host> can be the string "localhost" or the empty string; this is interpreted as `the machine from which the URL is being interpreted'.

Windows Explorer (at least) implements this rule as: `file://localhost/...` or `file:///...` turns into `...` as a DOS-type path, while any other `file://somehost/...` turns into a UNC path `//somehost/...`.

If that DOS-type path turns out to be a UNC path, it gets treated as a UNC path. I'm not sure if this is the URL handler, or the thing that the DOS-type path was passed to.

Anyway, this is why `file:///C:` and `file://localhost/C:` work, but `file://<myhostname>/C:` does not.

A shell sending OSC 7 per the terminal-wg spec, which specifies the Freedesktop file URL spec, will not generate `file://localhost/...`, because the Freedesktop file URL spec specifies the hostname in the `file://` URL should be either empty, or `gethostname()`. Unless of course `gethostbyname()` returns `localhost`, but that's hopefully very rare, and this isn't the only issue you'd see with that.

As mentioned, both Terminal.app and terminal-wg specify the file://<hostname>/ is intended for identifying remote sessions, so I would have expected (but have not checked) that shell implementations that generate OSC 7 commands would have been using `gethostbyname()` more often than not.

Anyway, how to _generate_ the URL is not the main issue for Windows Terminal, it's how to parse it. Ideally, we want to receive URLs that are already in the right format to just pass through `urlmon` or similar, and get back a UNC or filesystem path.

To get everything interoperating well, we need to handle (off hand):
* Paths from CMD, which are going to need to be `file://localhost/$CWD`, because CMD only knows DOS paths.
* Paths from cygwin-based systems, e.g., Git Bash or MSYS2), which I believe can only see DOS paths, and can convert from cygwin paths using `cygpath --windows ${PWD}`, and then it looks just like CMD, above.
* Paths from PowerShell, which could be DOS path _or_ UNC path. PowerShell can tell these apart, because `$executionContext.SessionState.Path.CurrentLocation` knows if it's on a 'Drive' or not, so this can easily do different things for either type of path.
* Paths from WSL, which from within the WSL session are Unix paths, and have two major mappings:
** `/mnt/X/...` where `X` is a drive-letter should map through to a DOS path for that Drive.
** Otherwise, it needs to to a UNC path `//wsl$/${WSL_DISTRO_NAME}/${CWD}`.
* Paths from an SSH session, which we ideally want to ignore early. However, these may look identical to a UNC path's `file://` URL. That might actually be desirable, if I am SSH'd to a machine on my LAN, and then launch a PowerShell tab and it does actually go into the same directory, via UNC. However, that'd be an unusual CIFs layout (NFS 3 worked that way though...) so I can't see this as being a _common_ use-case.

Happily, it seems `${WSL_DISTRO_NAME}` exists, and I assume it's inserted by the wsl shell launcher, so it is possible to generate the full correct `file://` URL in one's PS1 env-var. This even means you can key off the presence of `${WSL_DISTRO_NAME}` (or `${WSL_INTEROP}` perhaps) to share the same OSC 7 generation code with out-of-WSL environments.

One possible approach to the above: is that if everyone generating OSC 7 commands for Windows paths agrees to only generate a modified-'legacy file:// URL' format, i.e. `file://<gethostbyname()>///<UNC hostname>/<UNC path>` or `file://<gethostbyname()>/<DOS Path>`, then WT could recognise when `<gethostbyname()>` matches the current `$COMPUTERNAME`, and replace it with `localhost` (making it a correct "legacy" file:// URL) before passing it through to the URL handler to get the actual working directory.

Then any other hostname, except 'localhost' and 'empty string', could be treated as 'remote' and ignored. For 'localhost' and 'empty string' hosts, we'll just trust the user isn't generating those from a remote host. There's nothing else we can do there.

This seems the closest in-spirit to the Freedesktop file:// URL specification, and means that URLs generated for OSC 7 on Windows will be semantically-similar to URLs generated for OSC 7 on UNIX-type systems, and can be parsed the same way.

j4james commented:
> Paths from CMD, which are going to need to be `file://localhost/$CWD`, because CMD only knows DOS paths.

Ideally this should also have some form of hostname in place of localhost - I mentioned this in CMD gist.

> Paths from WSL, which from within the WSL session are Unix paths, and have two major mappings:
> * `/mnt/X/...` where `X` is a drive-letter should map through to a DOS path for that Drive.
> * Otherwise, it needs to to a UNC path `//wsl$/${WSL_DISTRO_NAME}/${CWD}`.

I'm not sure if I'm just misunderstanding you, but it sounds like you're saying the shell should be generating the path in one of those two formats, and that is just not an option. I've said this before, but I want to make sure this is understood. In any WSL shell we have to assume the URI will be in the form `file://$HOSTNAME$PWD`. If the user is in a `mnt/X` directory, then the URI will be `file://hostname/mnt/X`. If the user is in `/usr/bin`, the URI will be `file://hostname/usr/bin` - it's up to the Terminal to map that to the appropriate distro path for the session.

As for the UNC path format, I'm not sure why we're still arguing about that. Whatever you're producing in the PowerShell gist looks fine to me. I haven't looked at the git bash stuff because I know nothing about that - I'm going to assume that's fine too.

skyline75489 commented:
>The point of the host name is to distinguish between a path on a remote machine (which should be ignored) and a local one (which should be useable as the start path for the new tab).

In my initial implementation I was in the belief that based on the `PathFileExists` we can reliably filter local path and remote paths. And host name can not do more than that. So I chose to completely remove that part. Now after several days of discussion, I've been convinced that all things considered (including edge cases like PowerShell Remoting), host name does play a vital part in the whole OSC 7 implementation.

>Just to be clear, this transformation has to happen in the terminal itself, and not in the shell. 

Originally I dislike the idea of putting `/mnt/c` and `wsl$` related code inside WT, because they are super implementation-detailed. Putting them in WT harms the decoupling of WT and WSL and other shells. But it's also very tempting to make `/mnt/c` just work out of the box. So I'm stuck in the middle of nowhere.

TBBle commented:
@j4james: The problem is if the WSL session isn't generating paths in a Windows path layout, then we _must_ have special-case code in WT to detect a WSL session and mangle its URLs into the Windows path layout.

The only other option is we treat WSL as "remote", and just ignore its CWD.

_I_ would be fine with that (WSL is not a use-case for me here), but the same logic applies to git bash, and _I_ would definitely have my OSC 7 PS1 generate a _Windows-pathed_ `file://` URL using `cygpath --windows` in that case. I'd have to check, but wouldn't a vim build in git-bash be using Windows-based paths anyway, i.e. mingw but not msys? It's only actual shells that are likely to both manipulate the CWD and generate OSC 7, *and* be working in the cygwin path view.

There's no way outside of cygwin for WT to turn the cygwin `/bin` path into a Windows filesystem path, because that knowledge is local to the cygwin environment.

WSL is the same, except it's a mostly-fixed mapping, so we can _emulate_ it from the outside in WT. Like @skyline75489, I don't like that approach, but at least it's an MS product, so we could have a reasonably-fixed heuristic to recognise a WSL path, magically determine the WSL distribution from nothing, and then build the correct Windows-relative path.

> If the user is in a `mnt/X` directory, then the URI will be `file://hostname/mnt/X`. If the user is in `/usr/bin`, the URI will be `file://hostname/usr/bin` - it's up to the Terminal to map that to the appropriate distro path for the session.

I disagree about that. I don't think it's the terminal's responsibility to transform a path in a random filesystem-mapping wrapper into a Windows filesystem path. If the OSC 7-generating process in question cannot produce a Windows filesystem path, the _correct_ option is to ignore it. WT should not be expected to guess if `file://hostname/mnt/X` is a Cygwin drive mount of X:, a WSL drive mount of X:, or some random mount I happened to create inside WSL unrelated to a system drive.

skyline75489 commented:
I updated the PR and added hostname validation. I've tested more cases and it surprised me how well most shells behaves when accepting a Windows directory as starting directory. For example I can cd into `/etc` in WSL and Git Bash will happily accept it and change its current directory to `//wsl$/Ubuntu/etc`. It also works the other way around.

I think I'm close to the proper solution for this. Even if we are going to do the transformation for WSL, that would be a special case for WSL and for WSL only. Knowing the current behaviour of CMD, PowerShell and Git Bash makes me more convinced that all the other shells should be sending OSC 7 with a valid Windows path. 

> If the OSC 7-generating process in question cannot produce a Windows filesystem path, the correct option is to ignore it. 

I second this. I think this is the correct thing to do.

TBBle commented:
I did write "_correct_ option", but I meant "strictly correct". I don't think it's "incorrect" to have support for specific known sources of non-Windows-filesystem paths, but I would consider any such support separately to getting the OSC 7 feature working with `file://` URLs that _do_ reference the Windows filesystem path.

DHowett commented:
(Still working my way through the backlog...)

@TBBle 

> Edit: I kind-of wish WSL has something like `cygpath` now, to hide all the above

It does! WSL patches in a `wslpath` binary that does bidirectional path translation and takes into account `mtab`.

```
duhowett@dhowett-sl:~$ wslpath 'C:\Users'
/mnt/c/Users
duhowett@dhowett-sl:~$ wslpath '\\wsl$\Debian\home'
/home
duhowett@dhowett-sl:~$ wslpath -w /etc
\\wsl$\Debian\etc
duhowett@dhowett-sl:~$ wslpath -w /mnt/c
C:\
```

DHowett commented:
> we must have special-case code in WT to detect a WSL session and mangle its URLs into the Windows path layout.

This is, unfortunately, a feature request that comes up from time to time. People want the following things to "just work":

* Setting your `startingDirectory` to a WSL path
* Pasting a path off the clipboard (assuming we can even tell it's a path; do people think that this happens for free? :P)
* Dropping a file on the terminal window
* OSC 8 hyperlinks emitted by e.g. coreutils (see similar discussion in #7526)

But they can't _just work_ in the most general case. `wslpath` runs inside a specific distribution and looks into the mount table before responding.

I'd hate to go down the rabbit hole of attempting to determine if WSL happened to be running when a path was generated or a file was dropped or something. I just.. I don't have a good solution to this, in any form. Heuristics is a fancy word for "guessing your ass off"

skyline75489 commented:
I’m more than satisfied to know we have cygpath and wslpath at our disposal. This means we won’t need to “go down the rabbit hole of attempting to determine WSL is running”. We can still have a very smooth experience with the help of various shell confs in my gists above.

TBBle commented:
I agree, `cygpath` and `wslpath` remove the need to do any guessing, on the basis that the code _inside the environment_ can use those to get Windows-environment paths, so WSL can just consume the 'path' component of the `file://` URL without further mangling to account for the source environment.

skyline75489 commented:
Another interesting observation from me: `wslpath` is doing a fantastic job converting UNC paths. For a valid Linux path `my \folder`, the default tool `vte-urlencode-cwd` will convert it to `my%5Cfolder`, using the standard URL escape. However, `wslpath` will convert it to `myfolder`. The reason is obvious. The character `\` is not valid character for Windows paths. There must  be some kind of internal mapping inside WSL that handles this. And `wslpath` is fully aware of that.

![image](https://user-images.githubusercontent.com/4710575/93867604-dba16600-fcfb-11ea-8eff-02bb78671b51.png)


This is a great example showing that we actually *can not* use the path that a"normal" Linux would provide. If we want everything to work, the path conversion for WSL has to be handled inside the distro environment.

TBBle commented:
I was really hoping the substitute for [`\`](https://codepoints.net/U+005C) (0x5c) was going to be [`⧵`](https://codepoints.net/U+29F5), but it turns out it's [``](https://codepoints.net/U+F05C) (0xf05c), a private-use character. I like the alignment of the low-byte being the same.

It's a shame Windows Explorer doesn't appear to know how to render that character. I'm not sure how it _could_ but even so.

What happens if you use the character `` in a filename inside WSL? ^_^

PankajBhojwani commented on src/cascadia/TerminalCore/TerminalApi.cpp:
```diff
@@ -592,3 +594,55 @@ bool Terminal::EndHyperlink() noexcept
+    }
+
+    const auto prefix = uri.substr(0, 7);
+    if (!prefix.compare(L"file://") == 0)
```
The [Uri class](https://docs.microsoft.com/en-us/uwp/api/Windows.Foundation.Uri?view=winrt-19041) might be useful here - you can use it to easily obtain the scheme/hostname/path etc

skyline75489 replied:
I actually can not find a way to use `WIndows::Founcation::Uri` class. I have little experience with WinRT C++.

By the way do we need to worry about using this on Win7 or something?

TBBle replied:
Nope. Windows Terminal needs [Windows 10 1903 or later](https://github.com/microsoft/terminal#installing-and-running-windows-terminal).

TBBle replied:
The [WinRT introduction](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/intro-to-using-cpp-with-winrt#windows-desktop-application-cwinrt) has an example of specifically using `Windows.Foundation.Uri` from Win32 C++. It looks like the examples come with the WinRT VSIX extension.

The [C++ code itself looks reasonable](https://github.com/microsoft/cppwinrt/blob/master/vsix/ProjectTemplates/VC/Windows%20Desktop/ConsoleApplication/main.cpp), but if this is the first use of WinRT in the project, you might need to do other things to set up the include paths and linking. I've not done any development on Windows Terminal itself, so I don't know if that'll be simple like the [README implies](https://github.com/microsoft/cppwinrt/tree/master/vsix/ProjectTemplates/VC/Windows%20Desktop/ConsoleApplication), or complex if that README is hiding more complicated steps.

skyline75489 replied:
The shared components like DX renderer actually require Win7 support, because it will be integrated into Visual Studio in the future. I'm not sure `TerminalApi` falls into that category.

TBBle replied:
A quick poke around shows that [_TerminalCore/Terminal.cpp_](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalCore/Terminal.cpp) is already using WinRT, for access to `TerminalControl`, so I _expect_ it would be as easy as
````
#include <winrt/Windows.Foundation.h>
```

and then using `::winrt::Windows::Foundation::Uri` per [the docs](https://docs.microsoft.com/en-us/uwp/api/Windows.Foundation.Uri?view=winrt-18362). (Make sure the docs are in C++/WinRT mode, it defaults to JavaScript mode for me for some reason)

skyline75489 replied:
Despite the fact those WinRT goodies are very handy, I can not use it to parse the OSC 7 URI. The `::winrt::Windows::Foundation::Uri` class expect the URI format to be:

```plaintext
scheme://username:password @host:port/path.extension ?query #fragment
```

The FreeDesktop File URI spec is:

```plaintext
file://<hostname>/<path>
```

The behaviour of winrt Uri class is not very satisficing. For example for an Uri  `file:///D:/`, the winrt Uri gives you this:

```plaintext
Scheme: file
Host: 
Path: /D:/
```

Even weird, for an Uri in PowerShell `file://COMPUTERNAME/C:\\`, the winrt Uri class gives you this:

```plaintext
Scheme: file
Host: computername:
Path: /
```

I suspect that it can not handle `\` in a URI. And it prefers lowercase hostname? I don't really get it. We are, after all, parsing a FreeDesktop file URI so I can expect it to work. So that isn't a good choice to me.

TBBle replied:
Those URI parses examples are basically correct. I don't know what you're expecting differently in the first case? Hostnames in URIs are not case-sensitive (leads back to the DNS spec, which is not case-sensitive).

For the second case, `\` is not a valid path separator in a URI, so I somewhat expected a path `/C:\\/` but I'm not overly surprised that it just fails weirdly, since that's not a valid Windows file:// URI (I don't believe `:` is valid in a UNC path name) and I assume `Windows::Foundation::Uri` is noticing that. I know `system.uri` in PowerShell does checking for `file://` URIs and rejects "invalid" ones (like `file://wsl$/...`, which is invalid-per-the-URI-spec, even though it's valid (surprisingly) in the Windows file:// URI spec, because Windows supports non-DNS-valid hostnames).

TBBle replied:
That said, we should never see `\` in an OSC 7 URI, they need to be FreeDesktop File URI spec URIs (or as close as we can do) and the path separator is `/`, not `\`. So if we get `file:///D:/`, then the correct parse is as you gave above, which if we feed it back into urlmon, will be `D:\`.

skyline75489 replied:
Even though I get the correct path, the Uri class is not happy with the host name. For a URI `file://COMPUTER-NAME/D:/Projects`, the Uri class gives you this:

```plaintext
Scheme: file
Host: computername:
Path: /D:/Projects
```

Sounds like if we are using the Uri class we'll have to give up on hostname, because it is in fact not a valid file URI on WIndows. The hostname MUST be removed.

TBBle replied:
Interesting. Seems like a bug? `system.uri` handles that fine.
```
PS C:\Users\paulh> [system.uri]("file://COMPUTER-NAME/D:/Projects")

AbsolutePath   : /D:/Projects
AbsoluteUri    : file://computer-name/D:/Projects
LocalPath      : \\computer-name\D:\Projects
Authority      : computer-name
HostNameType   : Dns
IsDefaultPort  : True
IsFile         : True
IsLoopback     : False
PathAndQuery   : /D:/Projects
Segments       : {/, D:/, Projects}
IsUnc          : True
Host           : computer-name
Port           : -1
Query          :
Fragment       :
Scheme         : file
OriginalString : file://COMPUTER-NAME/D:/Projects
DnsSafeHost    : computer-name
IdnHost        : computer-name
IsAbsoluteUri  : True
UserEscaped    : False
UserInfo       :
```

DHowett commented:
Unfortunately, this component _must_ work on Windows 7. Anything “TerminalCore” and below (in the folders named conhost and shared in the SLN file) is Windows 7-clean, and anything in the rest of src/cascadia is Windows 10 1903.

TerminalCore builds into a component for Visual Studio. :)

DHowett commented:
The best thing to do would be to defer URI handling to TerminalControl or even higher.

skyline75489 commented:
@DHowett on the scale of 1-10, how much do you feel that I'm heading to the "correct" direction in this PR? I mean it may not be technically vital in the WT infrastructure but it's a popular request from the community.  I think we all want it to happen.

j4james commented:
I unsubscribed from this PR because it just makes me angry, but for the last time now, I'm *begging* you, if you can't implement this in a way that's compatible with the format used by existing shells and terminals, then please just invent another sequence for it. It's not going to make any difference to us if it's `OSC 7` or `OSC 4321` (or whatever), but we're just making problems for everyone else if we repurpose the existing sequence in an incompatible way.

TBBle commented:
Although @j4james has unsubscribed and won't see this, I want to say that as I have understood the discussion, the intention _is_ to implement OSC 7 as specified and used by existing terminals and shells: Shells (or others) send a `file://` URL, and if the Terminal can resolve it as local, then it is consider the CWD of that session, for whatever purposes the Terminal wants a CWD; in the WT case, for opening new tabs or sessions in the same CWD.

It would _not_ be compatible with other implementations for a WSL session to send `file://myhostname/usr/bin` and have Windows Terminal resolve that magically into the equivalent of `file://wsl$/MyWSLSession/usr/bin` or `file://localhost/wsl$/MyWSLSession/usr/bin`, in the same way that it would unexpected for a Linux desktop stack to resolve `file://myhostname/workspace` from inside a container into `file://myhosthome/var/lib/docker/volumes/SomeUUID/workspace`.

I specifically call out a container here because, like WSL and Cygwin, applications inside a container are running inside a mount namespace that is not visible or reasonably resolvable to things outside it. In theory, VTE could poke at the running processes, identify a container by interrogating the container runtime, parsing its volumes, and determining what the volume mapping was, if the CWD turns out to be a local filesystem mount.

WSL and Cygwin have the advantage that they provide a utility that _can_ produce a system-resolvable, `file://` path within their environment. That's up to the thing _generating_ the OSC 7 sequence to use, not the Terminal to try and parse it.

If the thing generating the OSC 7 sequence cannot produce a `file://` URL that works for the Terminal, it's not the Terminal's OSC 7 implementation that's faulty.

zadjii-msft commented:
Alright, I've re-read this thread and others, and I think I have an opinion on the matter.

Obviously, this is a super contentious sequence, which is why I hadn't already implemented it 😜. It's also an issue that's now intertwined with a couple other scenarios, so for the completeness of discussion, let's also refer to
* [#8166] `OSC 9;9` - the ConEmu sequence for setting the working directory
* [#7526] Support for the `file://` URI when emitted as part of `OSC 8` hyperlinks
* [#331] Users wanting to be able to drag/drop a file onto the Terminal running WSL, and have the pasted path be in WSL format, _not_ Windows
* [#592] Less importantly, but users also want to be able to set a
  `startingDirectory` as a WSL path, not just as a Windows path.


Basically, we've got two simultaneously incompatible issues. Either we:
* Support this sequence with _no_ "smart" path translation. We'll need to literally use whatever directory they emit (assuming the hostname matches)
    - Users who are currently using OSC 7 in their \*nix scripts and emitting `file://$hostname/$PWD` in WSL, cygwin, will get unexpected behavior. We'll try and treat `$PWD` as a path to a Windows directory - _which might be a path that exists_. When they try to open a new tab with that same directory, it'll open not in the distro the sequence originated in, but in the Windows filesystem. **Not great!**
* Find some way to magically translate paths provided by `OSC7` into the correct Windows path. WSL clients won't need to change any of the scripts they were already using for this to "just work".
    - There's no heuristic way of doing this, without the client application providing some extra information to us. We can't inspect the process tree to figure out if the process that's emitting this sequence is a WSL process or a docker process or an ssh, and even if we could, then we wouldn't be able to correlate `wsl.exe` to the distro within WSL that's being used. Similarly, we can't use the profile to do this, because the user might be using WSL w/in their "Command Prompt" profile (for example).
    - We could introduce some other sequence for additional metadata to be provided. Something like "I know I'm a WSL, and I'm `<distroName>`, when I emit paths, handle them specially". If we did introduce such a new custom sequence, then we're not _really_ faithfully implementing `OSC7` now are we? For clients to work as they did before, they'd still need to add emitting this custom sequence

I think at the end of the day, I agree with @j4james. If we implement `OSC7` as _requiring_ a Windows-style path, then that'll necessitate that client applications will need to be modified to work right on Windows, and I hate that idea.

#### `OSC9;9` - the ConEmu solution

This one doesn't seem that bad. Since it's a windows-first sequence, it doesn't need to know anything else about who's emitting the path. We can always assume it has been emitted as a Windows path. On the surface, this seems to work great, save for _one_ scenario - ssh. This sequence doesn't accept any sort of `hostname`, it only takes the path. So if you were `ssh`'d into another machine, and that one emitted a `OSC9;9`, then there's no way of determining that sequence wasn't intended for the current machine.

I think I'm okay with us implementing this as-is for the time being though. It's not a perfect sequence, but it's _good enough_, and doesn't come with the compat baggage that `OSC7` does. It can certainly be supported with the caveat "This won't work in an ssh session, or in any scenario where it's emitted by a remote connection. It will always be treated as a path on the local machine".

#### Other notes

I _want_ this to just work. I _want_ there to be one sequence that everyone can emit reliably, and have it _just work_, and I don't think that `OSC9;9` is that either. I'm gonna be filing a thread later today with the body of this post and a proposal for more discussion. This is a great feature that I think we all want, but we do need to make it work for both Windows and Linux terminals. The only thing I'm confident about in this whole discussion is that we (the terminal emulator ecosystem) _don't_ have a perfect solution yet.

skyline75489 commented:
Closed in favor of #8330
